src/tests/ffi.cpp

```diff
@@ -2084,9 +2084,391 @@ TEST_F(rnp_tests, test_ffi_key_generate_ex)
     assert_rnp_success(rnp_key_handle_destroy(subkey));
 
     /* Add SM2 subkey */
-    assert_rnp_success(rnp_op_generate_subkey_create(&keygen, ffi, key, "SM2"));
-    assert_rnp_success(rnp_op_generate_set_protection_cipher(keygen, "aes128"));
-    assert_rnp_success(rnp_op_generate_set_protection_hash(keygen, "sha1"));
+    assert_rnp_success(rnp_op_generate_subkey_create(&keygen, ffi, key, PGP_ALGNAME_SM2));
+    assert_rnp_success(rnp_op_generate_set_protection_cipher(keygen, PGP_ALGNAME_AES_128));
+    assert_rnp_success(rnp_op_generate_set_protection_hash(keygen, PGP_ALGNAME_SHA1));
+    assert_rnp_success(rnp_op_generate_execute(keygen));
+    assert_rnp_success(rnp_op_generate_get_key(keygen, &subkey));
+    assert_non_null(subkey);
+    assert_rnp_success(rnp_op_generate_destroy(keygen));
+    assert_rnp_success(rnp_key_handle_destroy(subkey));
+
+    assert_rnp_success(rnp_key_handle_destroy(key));
+    assert_rnp_success(rnp_ffi_destroy(ffi));
+}
+
+TEST_F(rnp_tests, test_ffi_key_generate_algnamecase)
```

Comment:
While purpose of this test is to check whether code works well with misc cases it is not needed to do additional calls - rnp_op_generate_set_bits(), _add_usage(), etc.
Also it would be much more clear to not use pgp_algname_rsa and so on but use direct constants - "Rsa", "aes128", etc. So if something failed then you'll easy what was the cause.